packages/core/test/bot-helpers/uniswap-broker/UniswapBrokerV3.js

```diff
@@ -0,0 +1,266 @@
+
+      await router.exactInput(params, { from: trader });
+
+      const deltaTokenA = tokenABefore.sub(await tokenA.balanceOf(trader));
+      const deltaTokenB = tokenBBefore.sub(await tokenB.balanceOf(trader));
+
+      // Token A should have increased by exactly 1 wei. This is the exact amount traded in the exactInput
+      assert.equal(deltaTokenA, toWei("1"));
+
+      // Token B should have decreased by the amount spent. This should be negative (tokens left the wallet) and should
+      // be about 0.01 due to the starting price in the pool. bound between 0 and -0.01.
+      assert.isTrue(deltaTokenB.gt(web3.utils.toWei("-0.01")) && deltaTokenB.lt(web3.utils.toWei("0")));
+    });
+  });
+
+  describe("Single liquidity provider", () => {
+    beforeEach(async () => {
+      // With just one liquidity provider, uniswapV3 acts like a standard AMM.
+      await addLiquidityToPool(toWei("1000"), toWei("100"), getTickFromPrice(8, fee), getTickFromPrice(15, fee));
+      // The starting price should be 10.
+      assert.equal((await getCurrentPrice(poolAddress)).toNumber(), 10);
+    });
+    it("Broker can correctly move the price up with a single liquidity provider", async function() {
+      // The broker should be able to trade up to a desired price. The starting price is 10. Try trade the market to 13.
+
+      await uniswapV3Broker.swapToPrice(
+        true, // Set Trading as EOA to true. This will pull tokens from the EOA and sent the output back to the EOA.
+        poolAddress, // Pool address for compting trade size.
+        router.address, // Router for executing the trade.
+        encodePriceSqrt(13, 1), // encoded target price of 13 defined as an X96 square root.
+        trader, // recipient of the trade.
+        MAX_UINT_VAL, // max deadline.
+        { from: trader }
+      );
+
+      // check the price moved up to 13 correctly.
+      const postTradePrice = (await getCurrentPrice(poolAddress)).toNumber();
+      assert.equal(postTradePrice, 13);
+    });
+
+    it("Broker can correctly move the price down with a single liquidity provider", async function() {
+      // The broker should be able to trade down to a desired price. The starting price at 10. Try trade the market to 8.5.
+
+      await uniswapV3Broker.swapToPrice(
+        true, // Set Trading as EOA to true. This will pull tokens from the EOA and sent the output back to the EOA.
+        poolAddress, // Pool address for compting trade size.
+        router.address, // Router for executing the trade.
+        encodePriceSqrt(8.5, 1), // encoded target price defined as an X96 square root.
+        trader, // recipient of the trade.
+        MAX_UINT_VAL, // max deadline.
+        { from: trader }
+      );
+
+      // check the price moved up to 12 correctly.
+      const postTradePrice = (await getCurrentPrice(poolAddress)).toNumber();
+      assert.equal(postTradePrice, 8.5);
+    });
+  });
+  describe("Multi-complex liquidity provider", () => {
```

Comment:
What happens if there is 0 liquidity for a certain price, but >0 liquidity below and above it? Does the Uniswap helper module account for that?